src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,652 @@
+        WithdrawalRequest memory request = withdrawalRequests[requestId];
+
+        require(request.claimTimestamp <= block.timestamp, "Claim delay not met");
+        // Is there enough liquidity to claim this request?
+        require(
+            request.queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this)),
+            "Queue pending liquidity"
+        );
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawsClaimed += request.assets;
+
+        assets = request.assets;
+
+        emit RedeemClaimed(msg.sender, requestId, assets);
+
+        // transfer the liquidity asset to the withdrawer
+        IERC20(liquidityAsset).transfer(msg.sender, assets);
+    }
+
+    /// @notice Check if a withdrawal request can be claimed
+    /// @param requestId The index of the withdrawal request
+    function isClaimable(uint256 requestId) public view returns (bool) {
+        return
+            withdrawalRequests[requestId].queued <= withdrawsClaimed + IERC20(liquidityAsset).balanceOf(address(this));
+    }
+
+    /// @dev Calculate how much of the liquidity asset (WETH) in the ARM is not reserved for the withdrawal queue.
+    // That is, the amount of liquidity assets (WETH) that is available to be swapped.
+    // If there are no outstanding withdrawals, then return the maximum uint256 value.
+    // The ARM can swap out liquidity assets (WETH) that has been accrued from the performance fee for the fee collector.
+    // There is no liquidity guarantee for the fee collector. If there is not enough liquidity assets (WETH) in
+    // the ARM to collect the accrued fees, then the fee collector will have to wait until there is enough liquidity assets.
+    function _liquidityAvailable() internal view returns (uint256) {
+        // The amount of WETH that is still to be claimed in the withdrawal queue
+        uint256 outstandingWithdrawals = withdrawsQueued - withdrawsClaimed;
+
+        // Save gas on an external balanceOf call if there are no outstanding withdrawals
+        if (outstandingWithdrawals == 0) return type(uint256).max;
+
+        // The amount of the liquidity asset is in the ARM
+        uint256 liquidityBalance = IERC20(liquidityAsset).balanceOf(address(this));
+
+        // If there is not enough liquidity assets in the ARM to cover the outstanding withdrawals
+        if (liquidityBalance <= outstandingWithdrawals) {
+            return 0;
+        }
+
+        return liquidityBalance - outstandingWithdrawals;
+    }
+
+    /// @notice The total amount of assets in the ARM and external withdrawal queue,
+    /// less the liquidity assets reserved for the ARM's withdrawal queue and accrued fees.
+    function totalAssets() public view virtual returns (uint256) {
```

Comment:
🟠 I'm concern that this method goes up and down in value as bi-driectional trades happen. This incentivizes cashing out at the top, reentering at bottom, and skipping out on the down parts.

Reply:
So here's the specific concern scenario. Assume we are operating in AMM mode. I'm using exaggerated price here for effect.

W -> STETH 0.90
STETH -> W 0.80

We'll assume the market is doing the trading, not the attacker.


- The market trades WETH for STETH. Because we offering a discount to get this off our hands, the assets we hold goes down by a bit.
- The attacker mints - it discounted by almost 10%, and they are getting an advantage over all other holders.
- The market trades STETH for WETH. They are selling at a discount, so the LP value goes up 20%.
- Attacker kicks off a redeem, locking in profits.

I don't think it's a question that this gets the attacker an unfair amount of yield. The primary next area of investigation is if this allows the attacker to steal principle from the other users. Another way of phrasing that is if the previewWithdraw ratio goes down as a result.

Reply:
My gut feel is fairly strong that this allows user to lose.

Reply:
The new `setPrice` logic ensures any stETH in the ARM can not be sold for less than what it was bought for. It does this by using a `crossPrice` that the price price must be below and the sell price must be equal or above.
Only the Owner can change the `crossPrice`, but they can not lower the `crossPrice` if there is stETH in the ARM contract.

Reply:
I've also changed `totalAssets` to price the base assets, eg stETH, at the new `crossPrice` instead of 1.0.

Here's and example 

Initial ARM state:

* sell price 0.9998
* cross price 0.9997
* but price 0.9996
* ARM has 10,000 WETH and no stETH

Actions:

1. ARM buys 10,004 stETH for 10,000 WETH at price 0.9996 stETH/WETH
 Total assets is 10,004 if stETH is valued at 1.0
  Total assets is 10,001 if stETH is valued at 0.9997

2. ARM sell 10,004 stETH for 10,002 WETH at price 0.9998 stETH/WETH
  Total assets now 10,002 as WETH is prices at 1.0

That's a 2 ETH profit from the start.
If stETH is priced at 1.0, the second action is a 2 ETH loss.
If stETH is priced at the cross price, 0.9997, the second action is a 1 ETH profit.

Reply:
I found a problem pricing the stETH in the Lido Withdrawal queue at the crossPrice. See the following example

Initial ARM state:

sell price 0.9998
cross price 0.9997
but price 0.9996
ARM has 10,000 WETH and no stETH

Actions:

1. ARM buys 10,004 stETH for 10,000 WETH at price 0.9996 stETH/WETH
  Total assets is 10,001 if stETH is priced at 0.9997

2. Operator puts the 10,004 in the withdrawal queue
  Total assets is 10,001 if stETH is priced at 0.9997

3. The Owner changes the prices to
  sell price 0.9995
  cross price 0.9994
  buy price 0.9993
  Total assets is 9,998 if stETH is priced at 0.9994

I think the stETH in the withdrawal queue needs to be priced at 1.0. So the new actions would be

2. Operator puts the 10,004 in the withdrawal queue
  Total assets is 10,004 if stETH in queue is priced at 1.0

3. The Owner changes the prices to
  sell price 0.9995
  cross price 0.9994
  buy price 0.9993
  Total assets is still 10,004 if stETH in queue is priced at 1.0